client/history/client.go

```diff
@@ -1020,6 +1015,121 @@ func (c *clientImpl) NotifyFailoverMarkers(
 	return nil
 }
 
+func (c *clientImpl) GetCrossClusterTasks(
+	ctx context.Context,
+	request *types.GetCrossClusterTasksRequest,
+	opts ...yarpc.CallOption,
+) (*types.GetCrossClusterTasksResponse, error) {
+	requestByClient := make(map[Client]*types.GetCrossClusterTasksRequest)
+	for _, shardID := range request.GetShardIDs() {
+		client, err := c.getClientForShardID(int(shardID))
+		if err != nil {
+			return nil, err
+		}
+
+		if _, ok := requestByClient[client]; !ok {
+			requestByClient[client] = &types.GetCrossClusterTasksRequest{
+				TargetCluster: request.TargetCluster,
+			}
+		}
+		requestByClient[client].ShardIDs = append(requestByClient[client].ShardIDs, shardID)
+	}
+
+	// preserve 5% timeout to return partial of the result if context is timing out
+	ctx, cancel := c.createChildContext(ctx, 0.05)
+	defer cancel()
+
+	futureByClient := make(map[Client]future.Future, len(requestByClient))
+	for client, req := range requestByClient {
+		future, settable := future.NewFuture()
+		go func() {
+			settable.Set(client.GetCrossClusterTasks(ctx, req))
+		}()
+
+		futureByClient[client] = future
+	}
+
+	response := &types.GetCrossClusterTasksResponse{
+		TasksByShard: make(map[int32][]*types.CrossClusterTaskRequest),
+	}
+	var err error
+	for _, future := range futureByClient {
+		var resp *types.GetCrossClusterTasksResponse
+		if futureErr := future.Get(ctx, &resp); futureErr != nil {
+			c.logger.Error("Failed to get cross cluster tasks", tag.Error(futureErr))
+			// TODO: return error for each shard and perform backoff at shard level.
+			// and ensure every shardID in request has a response (either tasks or failed cause).
+			//
+			// for _, failedShardID := range requestByClient[client].ShardIDs {
+			// 	response.FailedCauseByShard[failedShardID] = ...
+			// }
+			//
+			// for now following the pattern for getting replication tasks:
+			// ignore errors other than service busy, so that task fetcher in target
+			// cluster can slow down.
+			if common.IsServiceBusyError(futureErr) {
+				err = futureErr
```

Comment:
The behavior here is potentially to return a non-empty reponse and an error together since we aren't assigning nil to response when the error is set. If this is the intended behavior it would be good to put a comment on top of the `GetCrossClusterTasks` function to explain the contract. 

If you wanted to return nil when the error is set, then you can just say the for loop continues so we can make sure all the futures complete.

Reply:
It will wait until all futures complete since I am not breaking the for loop here I think. I just set the err value here when the error type is service busy. 

Yeah, so the plan for this diff is to return an error in this case. I will soon update the definition for the response struct to return per shard error as part of the response (as indicated by the TODO comment). Then we can return nil error and put error details in the response.